service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/shiro/realm/UserPassCredentialsMatcher.java

```diff
@@ -179,3 +199,42 @@ public boolean doCredentialsMatch(AuthenticationToken authenticationToken, Authe
     }
 
 }
+
+interface PasswordChecker {
+
+    boolean checkPassword(String tokenPassword, Credential infoCredential);
+
+}
+
+class CachePasswordChecker implements PasswordChecker {
+
+    private static final Cache<String, String> CACHED_CREDENTIALS = new LocalCache<String, String>(
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_SIZE, 1000),
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_TTL, 60),
+            null);
+
+    public boolean checkPassword(String tokenPassword, Credential infoCredential) {
+        String hashedTokenPassword = Base64.getEncoder().encodeToString(DigestUtils.sha3_512(tokenPassword));
+        String cachedCredential = CACHED_CREDENTIALS.get(hashedTokenPassword);
+        if (cachedCredential!=null && cachedCredential.equals(infoCredential.getCredentialKey())) {
```

Comment:
In the event of a hash collision (where the wrong password is provided but it produces the same hash as the correct password), I assume that `cachedCredential.equals(infoCredential.getCredentialKey())` should return false here?

Reply:
If the hash algorithm used is appropriate (anche sha3_512 should be) the likelihood of hash collision on wrong passwords should be minimal - particularly if the "wrong" password is very close to the correct one. That said, the cached credential seems to be the hashed password, so there is no way to distinguish it from a match against the correct one.

Reply:
Is the collision a possible event with bcrypt as well ?

Reply:
every hash algorithm has a collision risk because if you would like to have an n-bits output whethever input length you have if the possible inputs are m with m>2^n collisions are expected regardless of the algorithm you can find

Reply:
in other words if we have a password length of m bits and the hash produces n bits with n>m we could have no collisions (it depends on the hashing algorithm) otherwise there is no chance to have no collisions.